src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContextSerializer.cs

```diff
@@ -24,26 +29,156 @@ private DesigntimeLicenseContextSerializer()
-        internal static void Deserialize(Stream o, string cryptoKey, RuntimeLicenseContext context)
+        private class StreamWrapper : Stream
         {
-#pragma warning disable SYSLIB0011 // Issue https://github.com/dotnet/runtime/issues/39293 tracks finding an alternative to BinaryFormatter
-            IFormatter formatter = new BinaryFormatter();
+            private Stream _stream;
+            private bool _readFirstByte;
+            internal byte _firstByte;
+            public StreamWrapper(Stream stream)
+            {
+                _stream = stream;
+                _readFirstByte = false;
+                _firstByte = 0;
+            }
+
+            public override bool CanRead => _stream.CanRead;
+
+            public override bool CanSeek => _stream.CanSeek;
+
+            public override bool CanWrite => _stream.CanWrite;
+
+            public override long Length => _stream.Length;
+
+            public override long Position { get => _stream.Position; set => _stream.Position = value; }
+
+            public override void Flush() => _stream.Flush();
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                Debug.Assert(_stream.Position != 0, "Expected the first byte to be read first");
+                if (_stream.Position == 1)
+                {
+                    Debug.Assert(_readFirstByte == true);
+                    // Add the first byte read by ReadByte into buffer here
+                    buffer[offset] = _firstByte;
+                    return _stream.Read(buffer, offset + 1, count - 1) + 1;
+                }
+                return _stream.Read(buffer, offset, count);
+            }
+
+            public override long Seek(long offset, SeekOrigin origin) => _stream.Seek(offset, origin);
+
+            public override void SetLength(long value) => _stream.SetLength(value);
 
-            object obj = formatter.Deserialize(o);
+            public override void Write(byte[] buffer, int offset, int count) => _stream.Write(buffer, offset, count);
+
+            public override int ReadByte()
+            {
+                byte read = (byte)_stream.ReadByte();
+                _firstByte = read;
+                _readFirstByte = true;
+                return read;
+            }
+        }
+
+        /// <summary>
+        /// During deserialization, the stream passed in may be binary formatted or may have used binary writer. This is a quick test to discern between them.
+        /// </summary>
+        private static bool StreamIsBinaryFormatted(StreamWrapper stream)
+        {
+            // For binary formatter, the first byte is the SerializationHeaderRecord and has a value 0
+            int firstByte = stream.ReadByte();
+            if (firstByte != 0)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        [DynamicDependency(DynamicallyAccessedMemberTypes.NonPublicConstructors, typeof(Hashtable))]
```

Comment:
This means that the Hashtable constructors are going to be kept even when `EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization` is disabled. This is not linker friendly.

We generally do not care about linker compatibility with binary serialization enabled. Could you please delete this DynamicDependency if you agree?

Reply:
> This means that the Hashtable constructors are going to be kept even when EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization is disabled.

The Hashtable constructors will only be preserved if `EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization` is explicitly enabled. @pgovind is going to add a follow up PR in dotnet/sdk which will set `EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization=false` by default when `PublishTrimmed=true`.

This follows the same pattern with other feature switches that are set by default for trimmed apps:

* StartupHookSupport
* CustomResourceTypesSupport

See https://github.com/dotnet/sdk/blob/f922d2ef62dcaf6b99e64ed2ff48f772ad081d86/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets#L23-L41

Reply:
> If I remove the DynamicDependency here, an app with binary serialization enabled will fail at runtime because Reflection will try to construct a HashTable and fail to find the right constructor?

It is fine for an trimmed app with binary serialization enabled to fail at the runtime. If you enable BinarySerialization you will get a lot linker warnings about non-analyzable reflection uses. We do not plan to fix those warnings. BinarySerialization is deprecated feature and it is fine that it does not work with trimming. 
We have been deleting old code that was trying to make the binary serialization compatible with trimming, example https://github.com/mono/linker/pull/1925.

It is why I am suggesting to just delete this `[DynamicDependency]`. Deleting it would be consistent with approach we use everywhere else wrt binary serialization.

Reply:
I'm not sure I understand the concern here. If someone needs this in a trimmed app, they have to turn on the property to make it work. Why force them to also have to figure out which members need to be preserved so they can read their license keys (that may have been already saved in an assembly and they can't change them to use the new format)?

Reply:
Well "the library" in this case is our library - specifically System.Private.CoreLib and System.ComponentModel.TypeConverter. And we are fixing it to not use binary serialization (that's what this PR did).

I don't see a difference between "trimmed app" and "not trimmed app". The explicit goal of trimming is to make your app work the exact same way after trimming as it did before trimming.

If you are saying we should start throwing an exception here because it requires binary serialization, I think we should do it regardless of trimming.

Reply:
I am saying that we should be doing same thing here as what we are doing for `CustomResourceTypesSupport`. We do not make any effort to make `CustomResourceTypesSupport=true` to work well with the linker either.

If you turn on `CustomResourceTypesSupport`, you will get a ton a warning from the linker about bad use of reflection in binary serialization and the app probably not going to work.

If you turn on `EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization`, you will get a ton a warning from the linker about bad use of reflection in binary serialization and thus it is ok for the app to not work.

Reply:
> I am saying that we should be doing same thing here as what we are doing for CustomResourceTypesSupport.

I agree. And for `CustomResourceTypesSupport`, we are going to need to make it work in the context of WinForms apps. Or else WinForms apps that use resources (pretty much any real app) plainly won't work once trimmed. And WinForms, the owners of the scenario, will have to figure out the right things to preserve to make it work.

Here - we are the owners of the scenario. The runtime is calling this code. I don't think it is unreasonable for us to figure out what are the right things to preserve here - especially since it is so simple.

Reply:
> if we are not cleaning up all build-time linker warnings for given scenario

We did clean up the linker warnings for this scenario, because this was so simple - it is just (de)serialization `string` and `Hashtable`. And we added a trimming test to ensure it works.

Reply:
> And for CustomResourceTypesSupport, we are going to need to make it work in the context of WinForms apps.

And the scheme to make it work should not be based on binary serializer to deliver on https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md#net-6-nov-2021

> We did clean up the linker warnings for this scenario, because this was so simple

We should not be spending any time on improving binary serializer paths, even if it is simple. It is not a good use of time and it creates conflicting message about aggresively deprecating binary serializer.

Reply:
OK. I just felt that if we save a customer a few hours of trying to figure out why their app doesn't work, it would be worth it.

We can delete this attribute, plus the tests. and add the ILLink warnings to an `ILLink.Suppressions.LibraryBuild.xml` file.